File: structural/decorator/typescript/concrete-decorator.ts
import { NotifierDecorator } from './decorator';

export class NotifierTimerDecorator extends NotifierDecorator {
	async send(message: string) {
		const startTime = new Date().getTime();
		await super.send(message);
		const endTime = new Date().getTime();
		const elapsedTime = endTime - startTime;
		console.log(
			`[Design Pattern Decorator] Time taken to send notification: ${elapsedTime} milliseconds`
		);
	}
}

File: structural/decorator/python/concrete_decorator.py
import time

from decorator import NotifierDecorator


class NotifierTimerDecorator(NotifierDecorator):
    def send(self, message: str):
        start_time = time.time()
        self.notifier.send(message)
        end_time = time.time()
        elapsed_time = (end_time - start_time) * 1000
        print(
            "[Design Pattern Decorator] Time taken to send notification:"
            f" {elapsed_time:.2f} milliseconds"
        )
